pkg/cri/resource-manager/requests.go

```diff
@@ -189,6 +190,60 @@ func (m *resmgr) RunPod(ctx context.Context, method string, request interface{},
 	return reply, nil
 }
 
+// StopPod intercepts CRI requests for stopping Pods.
+func (m *resmgr) StopPod(ctx context.Context, method string, request interface{},
+	handler server.Handler) (interface{}, error) {
+
+	m.Lock()
+	defer m.Unlock()
+
+	podID := request.(*criapi.StopPodSandboxRequest).PodSandboxId
+	pod, ok := m.cache.LookupPod(podID)
+
+	if !ok {
+		m.Warn("%s: failed to look up pod %s, just passing request through", method, podID)
+	} else {
+		m.Info("%s: stopping pod %s (%s)...", method, pod.GetName(), podID)
+	}
+
+	reply, rqerr := handler(ctx, request)
+
+	if !ok {
+		return reply, rqerr
+	}
+
+	if rqerr != nil {
+		m.Error("%s: failed to stop pod %s: %v", method, podID, rqerr)
+	}
+
+	released := []cache.Container{}
+	for _, c := range pod.GetInitContainers() {
+		m.Info("%s: releasing resources for %s...", method, c.PrettyName())
+		if err := m.policy.ReleaseResources(c); err != nil {
+			m.Warn("%s: failed to release init-container %s: %v", method, c.PrettyName(), err)
+		}
+		c.UpdateState(cache.ContainerStateExited)
+		released = append(released, c)
+	}
+	for _, c := range pod.GetContainers() {
+		m.Info("%s: releasing resources for container %s...", method, c.PrettyName())
+		if err := m.policy.ReleaseResources(c); err != nil {
+			m.Warn("%s: failed to release container %s: %v", method, c.PrettyName(), err)
+		}
+		c.UpdateState(cache.ContainerStateExited)
+		released = append(released, c)
+	}
```

Comment:
Just wondered is it safe to release if the request fails or would it be better to wait for re-try? 🤔  E.g. the container runtime was not running at that very moment and didn't get the request. Kubelet tries again later at which point we've already lost track of the containers.

Reply:
Hmm... It's a valid question and I'm not sure which is a/the better way to go.

My approach was simply based on the observation that kubelet will not try to stop a pod unless it thinks all the containers in the pod are already stopped/have exited. Hence, it seemed/seems like a perfect synchronization point to make sure that we are of the same opinion.

Or, to look at it form an alternative perspective, we really don't know why the runtime failed to stop the pod. If it did because the pod does not exist any more, but both the kubelet and us still beg to differ, then ATM we have an effective allocation of resources for something which is not there. That allocation must eventually be (internally) released to prevent it from becoming a resource leak. And if the pod has multiple containers, we potentially have a bunch of those at this point.

I can change this to bail out on error without releasing the containers. Then (assuming that kubelet will not try to indefinitely stop/remove pods if those operations keep failing in the runtime) we have to use intercepted  `ListContainers()` with more complex logic than the current one (which simply releases containers that has recently exited). For some reason I just have a very much Matt Kowalski / Mardi Gras '87 feeling about that...

Reply:
I fully trust your judgement and reasoning here. Let's keep it the way you wrote it. Sounds safer to me that way. Anyway, my question was just something that occurred to me while reading the code

Reply:
Well, I already went ahead and updated it to not release containers on a failed pod stop request and added nuking to the processing of intercepted list request. Needs to bang that for a while to see if my worst fears are eased enough or if I need to roll back to the original simpler implementation.